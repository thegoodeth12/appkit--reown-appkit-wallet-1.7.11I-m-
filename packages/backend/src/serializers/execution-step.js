import stepSerializer from '@/serializers/step.js';

const executionStepSerializer = (executionStep) => {
  let executionStepData = {
    id: executionStep.id,
    dataIn: executionStep.dataIn,
    dataOut: executionStep.dataOut,
    errorDetails: executionStep.errorDetails,
    status: executionStep.status,
    createdAt: executionStep.createdAt.getTime(),
    updatedAt: executionStep.updatedAt.getTime(),
  };

  if (executionStep.step) {
    executionStepData.step = stepSerializer(executionStep.step);
  }

  return executionStepData;
};

export default executionStepSerializer;
