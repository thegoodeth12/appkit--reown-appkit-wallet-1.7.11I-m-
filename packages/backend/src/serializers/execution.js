import flowSerializer from '@/serializers/flow.js';

const executionSerializer = (execution) => {
  let executionData = {
    id: execution.id,
    testRun: execution.testRun,
    status: execution.status,
    createdAt: execution.createdAt.getTime(),
    updatedAt: execution.updatedAt.getTime(),
  };

  if (execution.status) {
    executionData.status = execution.status;
  }

  if (execution.flow) {
    executionData.flow = flowSerializer(execution.flow);
  }

  return executionData;
};

export default executionSerializer;
