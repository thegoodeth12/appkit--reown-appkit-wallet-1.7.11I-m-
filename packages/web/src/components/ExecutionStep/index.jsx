import PropTypes from 'prop-types';
import * as React from 'react';
import { DateTime } from 'luxon';
import Stack from '@mui/material/Stack';
import ErrorIcon from '@mui/icons-material/Error';
import CheckCircleIcon from '@mui/icons-material/CheckCircle';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Typography from '@mui/material/Typography';
import Tooltip from '@mui/material/Tooltip';
import Box from '@mui/material/Box';
import Chip from '@mui/material/Chip';

import TabPanel from 'components/TabPanel';
import SearchableJSONViewer from 'components/SearchableJSONViewer';
import AppIcon from 'components/AppIcon';
import useFormatMessage from 'hooks/useFormatMessage';
import useApps from 'hooks/useApps';
import {
  AppIconWrapper,
  AppIconStatusIconWrapper,
  Content,
  Header,
  Metadata,
  Wrapper,
} from './style';
import { ExecutionStepPropType, StepPropType } from 'propTypes/propTypes';

function ExecutionStepId(props) {
  const formatMessage = useFormatMessage();

  const id = (
    <Typography variant="caption" component="span">
      {props.id}
    </Typography>
  );

  return (
    <Box sx={{ display: 'flex' }} gridArea="id">
      <Typography
        data-test="execution-step-id"
        variant="caption"
        fontWeight="bold"
      >
        {formatMessage('executionStep.id', { id })}
      </Typography>
    </Box>
  );
}

ExecutionStepId.propTypes = {
  id: PropTypes.string.isRequired,
};

function ExecutionStepDate(props) {
  const formatMessage = useFormatMessage();
  const createdAt = DateTime.fromMillis(parseInt(props.createdAt, 10));
  const relativeCreatedAt = createdAt.toRelative();

  return (
    <Tooltip
      title={createdAt.toLocaleString(DateTime.DATETIME_FULL_WITH_SECONDS)}
    >
      <Typography
        data-test="execution-step-executed-at"
        variant="caption"
        gutterBottom
      >
        {formatMessage('executionStep.executedAt', {
          datetime: relativeCreatedAt,
        })}
      </Typography>
    </Tooltip>
  );
}

ExecutionStepDate.propTypes = {
  createdAt: PropTypes.number.isRequired,
};

const validIcon = <CheckCircleIcon color="success" />;
const errorIcon = <ErrorIcon color="error" />;

function ExecutionStep(props) {
  const { executionStep } = props;
  const [activeTabIndex, setActiveTabIndex] = React.useState(0);
  const step = executionStep.step;
  const isTrigger = step.type === 'trigger';
  const isAction = step.type === 'action';
  const formatMessage = useFormatMessage();
  const useAppsOptions = {};

  if (isTrigger) {
    useAppsOptions.onlyWithTriggers = true;
  }

  if (isAction) {
    useAppsOptions.onlyWithActions = true;
  }

  const { data: apps } = useApps(useAppsOptions);

  const app = apps?.data?.find((currentApp) => currentApp.key === step.appKey);

  if (!apps?.data) return null;

  const validationStatusIcon =
    executionStep.status === 'success' ? validIcon : errorIcon;

  const hasError = !!executionStep.errorDetails;

  const stepTypeName = isTrigger
    ? formatMessage('flowStep.triggerType')
    : formatMessage('flowStep.actionType');

  return (
    <Wrapper elevation={1} data-test="execution-step">
      <Header>
        <Stack direction="row" gap={3}>
          <AppIconWrapper>
            <AppIconStatusIconWrapper>
              <AppIcon
                url={app?.iconUrl}
                name={app?.name}
                color={app?.primaryColor}
              />
              {validationStatusIcon}
            </AppIconStatusIconWrapper>
          </AppIconWrapper>

          <Metadata flex="1">
            <ExecutionStepId id={executionStep.step.id} />

            <Box flex="1" gridArea="step">
              <Typography
                component={Stack}
                direction="row"
                variant="stepApp"
                alignItems="center"
                gap={0.5}
              >
                <Chip
                  data-test="step-type"
                  label={stepTypeName}
                  variant="stepType"
                  size="small"
                />

                {app?.name}
              </Typography>

              <Typography data-test="step-position-and-name" variant="body2">
                {step.position}. {step.name}
              </Typography>
            </Box>

            <Box
              display="flex"
              justifyContent={['left', 'right']}
              gridArea="date"
            >
              <ExecutionStepDate createdAt={executionStep.createdAt} />
            </Box>
          </Metadata>
        </Stack>
      </Header>

      <Content sx={{ px: 2 }}>
        <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
          <Tabs
            value={activeTabIndex}
            onChange={(event, tabIndex) => setActiveTabIndex(tabIndex)}
          >
            <Tab label="Data in" data-test="data-in-tab" />
            <Tab label="Data out" data-test="data-out-tab" />
            {hasError && <Tab label="Error" data-test="error-tab" />}
          </Tabs>
        </Box>

        <TabPanel value={activeTabIndex} index={0} data-test="data-in-panel">
          <SearchableJSONViewer data={executionStep.dataIn} />
        </TabPanel>

        <TabPanel value={activeTabIndex} index={1} data-test="data-out-panel">
          <SearchableJSONViewer data={executionStep.dataOut} />
        </TabPanel>

        {hasError && (
          <TabPanel value={activeTabIndex} index={2}>
            <SearchableJSONViewer data={executionStep.errorDetails} />
          </TabPanel>
        )}
      </Content>
    </Wrapper>
  );
}

ExecutionStep.propTypes = {
  collapsed: PropTypes.bool,
  step: StepPropType.isRequired,
  index: PropTypes.number,
  executionStep: ExecutionStepPropType.isRequired,
};

export default ExecutionStep;
