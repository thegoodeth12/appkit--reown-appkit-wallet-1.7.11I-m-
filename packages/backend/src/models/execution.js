import { DateTime } from 'luxon';
import Base from '@/models/base.js';
import Flow from '@/models/flow.js';
import ExecutionStep from '@/models/execution-step.js';
import Telemetry from '@/helpers/telemetry/index.js';

class Execution extends Base {
  static tableName = 'executions';

  static jsonSchema = {
    type: 'object',

    properties: {
      id: { type: 'string', format: 'uuid' },
      flowId: { type: 'string', format: 'uuid' },
      testRun: { type: 'boolean', default: false },
      internalId: { type: 'string' },
      status: { type: 'string', enum: ['success', 'failure'] },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    flow: {
      relation: Base.BelongsToOneRelation,
      modelClass: Flow,
      join: {
        from: 'executions.flow_id',
        to: 'flows.id',
      },
    },
    executionSteps: {
      relation: Base.HasManyRelation,
      modelClass: ExecutionStep,
      join: {
        from: 'executions.id',
        to: 'execution_steps.execution_id',
      },
    },
  });

  static find({ name, status, startDateTime, endDateTime }) {
    return this.query()
      .withSoftDeleted()
      .joinRelated({
        flow: true,
      })
      .withGraphFetched({
        flow: {
          steps: true,
        },
      })
      .where((builder) => {
        builder.withSoftDeleted();

        if (name) {
          builder.where('flow.name', 'ilike', `%${name}%`);
        }

        if (status === 'success') {
          builder.where('executions.status', 'success');
        } else if (status === 'failure') {
          builder.where('executions.status', 'failure');
        }

        if (startDateTime) {
          const startDate = DateTime.fromMillis(Number(startDateTime));

          if (startDate.isValid) {
            builder.where('executions.created_at', '>=', startDate.toISO());
          }
        }

        if (endDateTime) {
          const endDate = DateTime.fromMillis(Number(endDateTime));

          if (endDate.isValid) {
            builder.where('executions.created_at', '<=', endDate.toISO());
          }
        }
      })
      .orderBy('created_at', 'desc');
  }

  async $afterInsert(queryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.executionCreated(this);
  }
}

export default Execution;
