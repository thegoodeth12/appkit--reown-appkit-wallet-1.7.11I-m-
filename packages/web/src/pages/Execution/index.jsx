import * as React from 'react';
import { useParams } from 'react-router-dom';
import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import AlertTitle from '@mui/material/AlertTitle';
import Alert from '@mui/material/Alert';

import useFormatMessage from 'hooks/useFormatMessage';
import ExecutionHeader from 'components/ExecutionHeader';
import ExecutionStep from 'components/ExecutionStep';
import Container from 'components/Container';
import useExecutionSteps from 'hooks/useExecutionSteps';
import useExecution from 'hooks/useExecution';

export default function Execution() {
  const { executionId } = useParams();
  const formatMessage = useFormatMessage();

  const { data: execution } = useExecution({ executionId });

  const {
    data,
    isLoading: isExecutionStepsLoading,
    fetchNextPage,
    hasNextPage,
    isFetching,
    isFetchingNextPage,
  } = useExecutionSteps({
    executionId: executionId,
  });

  React.useEffect(() => {
    if (!isFetching && !isFetchingNextPage && hasNextPage) {
      fetchNextPage();
    }
  }, [isFetching, isFetchingNextPage, hasNextPage, fetchNextPage]);

  return (
    <Container sx={{ py: 3 }}>
      <ExecutionHeader execution={execution?.data} />

      <Grid container item sx={{ mt: 2, mb: [2, 5] }} rowGap={3}>
        {!isExecutionStepsLoading && !data?.pages?.[0].data.length && (
          <Alert severity="warning" sx={{ flex: 1 }}>
            <AlertTitle>{formatMessage('execution.noDataTitle')}</AlertTitle>

            <Box>{formatMessage('execution.noDataMessage')}</Box>
          </Alert>
        )}

        {data?.pages?.map((group, i) => (
          <React.Fragment key={i}>
            {group?.data?.map((executionStep) => (
              <ExecutionStep
                key={executionStep.id}
                executionStep={executionStep}
                step={executionStep.step}
              />
            ))}
          </React.Fragment>
        ))}
      </Grid>
    </Container>
  );
}
