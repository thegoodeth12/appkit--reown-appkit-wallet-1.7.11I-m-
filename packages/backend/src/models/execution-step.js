import appConfig from '@/config/app.js';
import Base from '@/models/base.js';
import Execution from '@/models/execution.js';
import Step from '@/models/step.js';
import Telemetry from '@/helpers/telemetry/index.js';

class ExecutionStep extends Base {
  static tableName = 'execution_steps';

  static jsonSchema = {
    type: 'object',

    properties: {
      id: { type: 'string', format: 'uuid' },
      executionId: { type: 'string', format: 'uuid' },
      stepId: { type: 'string' },
      dataIn: { type: ['object', 'null'] },
      dataOut: { type: ['object', 'null'] },
      status: { type: 'string', enum: ['success', 'failure'] },
      errorDetails: { type: ['object', 'null'] },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    execution: {
      relation: Base.BelongsToOneRelation,
      modelClass: Execution,
      join: {
        from: 'execution_steps.execution_id',
        to: 'executions.id',
      },
    },
    step: {
      relation: Base.BelongsToOneRelation,
      modelClass: Step,
      join: {
        from: 'execution_steps.step_id',
        to: 'steps.id',
      },
    },
  });

  get isFailed() {
    return this.status === 'failure';
  }

  async isSucceededNonTestRun() {
    const execution = await this.$relatedQuery('execution');
    return !execution.testRun && !this.isFailed;
  }

  async updateUsageData() {
    const execution = await this.$relatedQuery('execution');

    const flow = await execution.$relatedQuery('flow');
    const user = await flow.$relatedQuery('user');
    const usageData = await user.$relatedQuery('currentUsageData');

    await usageData.increaseConsumedTaskCountByOne();
  }

  async increaseUsageCount() {
    if (appConfig.isCloud && this.isSucceededNonTestRun()) {
      await this.updateUsageData();
    }
  }

  async updateExecutionStatus() {
    const execution = await this.$relatedQuery('execution');

    await execution.$query().patch({
      status: this.status === 'failure' ? 'failure' : 'success',
    });
  }

  async $afterInsert(queryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.executionStepCreated(this);
    await this.increaseUsageCount();
    await this.updateExecutionStatus();
  }
}

export default ExecutionStep;
