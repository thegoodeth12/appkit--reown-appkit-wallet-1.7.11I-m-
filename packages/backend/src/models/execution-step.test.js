import { vi, describe, it, expect } from 'vitest';
import Execution from '@/models/execution.js';
import ExecutionStep from '@/models/execution-step.js';
import Step from '@/models/step.js';
import Base from '@/models/base.js';
import UsageData from '@/models/usage-data.ee.js';
import Telemetry from '@/helpers/telemetry/index.js';
import appConfig from '@/config/app.js';
import { createExecution } from '@/factories/execution.js';
import { createExecutionStep } from '@/factories/execution-step.js';

describe('ExecutionStep model', () => {
  it('tableName should return correct name', () => {
    expect(ExecutionStep.tableName).toBe('execution_steps');
  });

  it('jsonSchema should have correct validations', () => {
    expect(ExecutionStep.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = ExecutionStep.relationMappings();

    const expectedRelations = {
      execution: {
        relation: Base.BelongsToOneRelation,
        modelClass: Execution,
        join: {
          from: 'execution_steps.execution_id',
          to: 'executions.id',
        },
      },
      step: {
        relation: Base.BelongsToOneRelation,
        modelClass: Step,
        join: {
          from: 'execution_steps.step_id',
          to: 'steps.id',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  describe('isFailed', () => {
    it('should return true if status is failure', async () => {
      const executionStep = new ExecutionStep();
      executionStep.status = 'failure';

      expect(executionStep.isFailed).toBe(true);
    });

    it('should return false if status is not failure', async () => {
      const executionStep = new ExecutionStep();
      executionStep.status = 'success';

      expect(executionStep.isFailed).toBe(false);
    });
  });

  describe('isSucceededNonTestRun', () => {
    it('should return false if it has a test run execution', async () => {
      const execution = await createExecution({
        testRun: true,
      });

      const executionStep = await createExecutionStep({
        executionId: execution.id,
      });

      expect(await executionStep.isSucceededNonTestRun()).toBe(false);
    });

    it('should return false if it has a failure status', async () => {
      const executionStep = await createExecutionStep({
        status: 'failure',
      });

      expect(await executionStep.isSucceededNonTestRun()).toBe(false);
    });

    it('should return true if it has a succeeded non test run', async () => {
      const executionStep = await createExecutionStep({
        status: 'success',
      });

      expect(await executionStep.isSucceededNonTestRun()).toBe(true);
    });
  });

  describe('updateUsageData', () => {
    it('should call usageData.increaseConsumedTaskCountByOne', async () => {
      const executionStep = await createExecutionStep();

      const increaseConsumedTaskCountByOneSpy = vi.spyOn(
        UsageData.prototype,
        'increaseConsumedTaskCountByOne'
      );

      await executionStep.updateUsageData();

      expect(increaseConsumedTaskCountByOneSpy).toHaveBeenCalledOnce();
    });
  });

  describe('increaseUsageCount', () => {
    it('should call updateUsageData for cloud and succeeded non test run', async () => {
      vi.spyOn(appConfig, 'isCloud', 'get').mockReturnValue(true);
      vi.spyOn(
        ExecutionStep.prototype,
        'isSucceededNonTestRun'
      ).mockReturnValue(true);

      const executionStep = await createExecutionStep();

      const updateUsageDataSpy = vi.spyOn(
        ExecutionStep.prototype,
        'updateUsageData'
      );

      await executionStep.increaseUsageCount();

      expect(updateUsageDataSpy).toHaveBeenCalledOnce();
    });
  });

  describe('$afterInsert', () => {
    it('should call Telemetry.executionStepCreated', async () => {
      const telemetryExecutionStepCreatedSpy = vi
        .spyOn(Telemetry, 'executionStepCreated')
        .mockImplementation(() => {});

      const executionStep = await createExecutionStep();

      expect(telemetryExecutionStepCreatedSpy).toHaveBeenCalledWith(
        executionStep
      );
    });

    it('should call increaseUsageCount', async () => {
      const increaseUsageCountSpy = vi.spyOn(
        ExecutionStep.prototype,
        'increaseUsageCount'
      );

      await createExecutionStep();

      expect(increaseUsageCountSpy).toHaveBeenCalledOnce();
    });
  });

  describe('updateExecutionStatus', () => {
    it('should update execution status to failure when step status is failure', async () => {
      const execution = await createExecution();
      const executionStep = await createExecutionStep({
        executionId: execution.id,
        status: 'failure',
      });

      await executionStep.updateExecutionStatus();

      const updatedExecution = await execution.$query();
      expect(updatedExecution.status).toBe('failure');
    });

    it('should update execution status to success when step status is success', async () => {
      const execution = await createExecution();
      const executionStep = await createExecutionStep({
        executionId: execution.id,
        status: 'success',
      });

      await executionStep.updateExecutionStatus();

      const updatedExecution = await execution.$query();
      expect(updatedExecution.status).toBe('success');
    });
  });
});
