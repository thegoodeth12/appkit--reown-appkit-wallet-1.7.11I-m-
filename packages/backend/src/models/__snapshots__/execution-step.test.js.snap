// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExecutionStep model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "createdAt": {
      "type": "string",
    },
    "dataIn": {
      "type": [
        "object",
        "null",
      ],
    },
    "dataOut": {
      "type": [
        "object",
        "null",
      ],
    },
    "deletedAt": {
      "type": "string",
    },
    "errorDetails": {
      "type": [
        "object",
        "null",
      ],
    },
    "executionId": {
      "format": "uuid",
      "type": "string",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "status": {
      "enum": [
        "success",
        "failure",
      ],
      "type": "string",
    },
    "stepId": {
      "type": "string",
    },
    "updatedAt": {
      "type": "string",
    },
  },
  "type": "object",
}
`;
